### ISSUE TITLE
Cold Slate Copperhead - Scale calculation rounds in favor of user during withdraw

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-01-napier-judging#56]
Cold Slate Copperhead

high

# Scale calculation rounds in favor of user during withdraw

## Summary

The withdrawal "divides by the ERC4626 `convertToAssets`" which results in a subtle round-in-favor-of-user.

## Vulnerability Detail

Multiplying shares `convertToAssets` in an ERC4626 vault converts from shares to assets. However, dividing by `convertToAssets` should not be used for converting the other way around (shares -> assets) because since the divisor rounds down this ends up as the conversion "rounding up".

`scale` is essentially the ERC4626 `convertToAssets` function with `1e18` as an input parameter. In this equation, the `maxscale` which was derived from `scale` is used as a divisor which has the "round up problem"

```solidity
uint256 sharesRedeemed = pyAmount.divWadDown(_gscales.maxscale);
```

Important note: this issue does not refer to the rounding of `divWadDown` which correctly rounds down. It refers to the rounding down of `maxscale/scale` which converts to a round up since sharesRedeemed is inversely proportional to scale.

`maxScale` was initially based off `convertToAssets` which we know rounds down. Since `sharesRedeemed` is inversely proportional to `maxScale` the rounding down of `maxScale` converts to a rounding up in sharesRedeemed. This can result in a multiple small withdrawals giving more shares back than the initial deposit even though no yield was accrued.

**Using share inflation to exacerbate the impact**

We can combine this rounding-in-favor with a share inflation during the initial deposit. An attacker can inflate the shares during the initial deposit to $10 USD per share through token donation. This amount is selected so the rounding error may be small enough to not be noticed by regular depositors but is large enough that the favorable rounding exceeds gas costs. Then, they could call withdraw with the round-in-favor repeatedly to drain the entire tranche.

## Impact

In normal use, this could result in the last withdrawer having their withdrawal revert due to slight insolvency of the protocol due to rounding in favor of the user. Since the user must attempt to withdraw their entire balance rather than entering an amount parameter in `collect()`, this can result in a significant amount of stuck funds.

By combining this vulnerability with inflated shares, the attacker could drain the protocol with many withdrawals with favorable rounding.

## Code Snippet

https://github.com/sherlock-audit/2024-01-napier/blob/main/napier-v1/src/Tranche.sol#L219

## Tool used

Manual Review

## Recommendation

Add +1 to the denominator. This basically converts the `maxscale` to a rounded up version.

```solidity
uint256 sharesRedeemed = pyAmount.divWadDown(_gscales.maxscale + 1);
```

**Comment by Banditx0x** [comment]
Escalate. 

This issue is actually different from the duplicated issue #96 .  This makes a unique claim that the rounding direction is infact incorrect, which is more severe than the duplicated issue claims: 

As stated in issue 96, the claimed issue is a rounding down leading to ERC non-compliance.

_"The main problem here is that the division done here is rounded down. Thus, the sharesRedeem will be lower than expected. "_

My issue actually claims that the rounding DIRECTION is wrong.

In this issue (#56), the rounding _within_ `cScale` and this `maxScale` rounds DOWN. Since `maxScale` is the denominator in the equation `uint256 sharesRedeemed = pyAmount.divWadDown(_gscales.maxscale);` the `sharesRedeemed` rounds UP. The `ERC5095` issue and all its duplicates makes no such claim.

**Summary:**

- `cScale` which is used to compute `maxScale` rounds down as it is queried off `convertSharesToAssets` (rounds down).
- since `sharesRedeemed` is inversely proportional to `maxScale` (`maxScale` is divisor). The rounding down when computing `maxScale` means that the divisor can be LOWER, which means that the result `sharesRedeemed` is HIGHER.

**Comment by sherlock-admin2** [comment]
> Escalate. 
> 
> This issue is actually different from the duplicated issue #96 .  This makes a unique claim that the rounding direction is infact incorrect, which is more severe than the duplicated issue claims: 
> 
> As stated in issue 96, the claimed issue is a rounding down leading to ERC non-compliance.
> 
> _"The main problem here is that the division done here is rounded down. Thus, the sharesRedeem will be lower than expected. "_
> 
> My issue actually claims that the rounding DIRECTION is wrong.
> 
> In this issue (#56), the rounding _within_ `cScale` and this `maxScale` rounds DOWN. Since `maxScale` is the denominator in the equation `uint256 sharesRedeemed = pyAmount.divWadDown(_gscales.maxscale);` the `sharesRedeemed` rounds UP. The `ERC5095` issue and all its duplicates makes no such claim.
> 
> **Summary:**
> 
> - `cScale` which is used to compute `maxScale` rounds down as it is queried off `convertSharesToAssets` (rounds down).
> - since `sharesRedeemed` is inversely proportional to `maxScale` (`maxScale` is divisor). The rounding down when computing `maxScale` means that the divisor can be LOWER, which means that the result `sharesRedeemed` is HIGHER.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by cvetanovv** [comment]
I agree with @Banditx0x escalation. We can separate it as a unique Medium.

**Comment by nevillehuang** [comment]
I believe This issues impact and root cause should be considered alongside #16

**Comment by cvetanovv** [comment]
I think they should stay separate. I agree that the root is the same: "rounding error". From there on, the descriptions of the two reports are different. Also, the attack vectors that are shown by the Watsons are different.

**Comment by Czar102** [comment]
@nevillehuang given that there was no escalation to make this a duplicate of #16, the fact that you have not elaborated on the reasons and that I don't see an immediate similarity, I'm not planning to consider these duplicates.

Planning to accept the escalation and consider this submission a separate Medium. Are there any duplicates of this finding @cvetanovv?

**Comment by nevillehuang** [comment]
@Czar102 The impact is similar to #16, but yea since nobody escalated including me with this logic up to you

**Comment by nevillehuang** [comment]
@massun-onibakuchi Can you confirm the same workaround mentioned [here](https://github.com/sherlock-audit/2024-01-napier-judging/issues/16#issuecomment-1985628392) applies albeit the rounding is initiated from different causes?

**Comment by massun-onibakuchi** [comment]
> @massun-onibakuchi Can you confirm the same workaround mentioned [here](https://github.com/sherlock-audit/2024-01-napier-judging/issues/16#issuecomment-1985628392) applies albeit the rounding is initiated from different causes?

@nevillehuang 
Sorry for missing the mention to me. I think the workaround would work in this case.

**Comment by massun-onibakuchi** [comment]
When a last user is PT holder, he/she can avoid revert relatively easily. When a last user is YT holder, he/she may need the workaround above.

**Comment by Czar102** [comment]
I see, yes. It seems #16 is invalidated anyway. Planning to accept the escalation and invalidate this report (and make it a duplicate of #16).

**Comment by BanditSecurity** [comment]
@Czar102 Why would this be invalidated due to #16 ? These are different issues

**Comment by Czar102** [comment]
@BanditSecurity @cvetanovv can you elaborate on why is this issue different from #16?

**Comment by cvetanovv** [comment]
I initially thought #16 and this issue was valid. But after the explanation of the sponsor's comment and given that #16 is already low, I think 56 should also be low because the impact and behavior are the same.

**Comment by Czar102** [comment]
Result:
Low
Unique

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [banditx0x](https://github.com/sherlock-audit/2024-01-napier-judging/issues/56/#issuecomment-1985544965): accepted
